compiler/src/dotty/tools/backend/sjs/JUnitBootstrappers.scala

```diff
@@ -0,0 +1,251 @@
+package dotty.tools.backend.sjs
+
+import scala.annotation.tailrec
+
+import dotty.tools.dotc._
+
+import dotty.tools.dotc.core._
+import Constants._
+import Contexts._
+import Decorators._
+import Flags._
+import Names._
+import NameOps._
+import Phases._
+import Scopes._
+import Symbols._
+import StdNames._
+import Types._
+
+import dotty.tools.dotc.transform.MegaPhase._
+
+/** Generates JUnit bootstrapper classes for Scala.js. */
+class JUnitBootstrappers extends MiniPhase {
+  import ast.tpd._
+
+  def phaseName: String = "junitBootstrappers"
+
+  override def isEnabled(implicit ctx: Context): Boolean =
+    super.isEnabled && ctx.settings.scalajs.value
+
+  private object Names {
+    val beforeClass: TermName = termName("beforeClass")
+    val afterClass: TermName = termName("afterClass")
+    val before: TermName = termName("before")
+    val after: TermName = termName("after")
+    val tests: TermName = termName("tests")
+    val invokeTest: TermName = termName("invokeTest")
+    val newInstance: TermName = termName("newInstance")
+
+    val instance: TermName = termName("instance")
+    val name: TermName = termName("name")
+  }
+
+  private class MyDefinitions()(implicit ctx: Context) {
```

Comment:
Can this stuff go into JSDefinitions instead ?

Reply:
Right now it could, but this phase really should go into a separate compiler plugin at some point. For scalac it is in its own nsc plugin. The reason is that you only want to enable it on a) the *test* configuration and b) only if you actually use JUnit tests; otherwise, it's a lot of work done for nothing.

So basically I wrote it in the compiler right now, but we should extract it in a compiler plugin at some point, and therefore I wrote it in a way that does not invade the rest of the codebase.

Reply:
Well, the compiler already has to handle JUnit tests in an earlier phase: https://github.com/lampepfl/dotty/blob/1c3b45a3d292fb97a3316cb3e6b118b2966d4d95/compiler/src/dotty/tools/dotc/transform/MixinOps.scala#L66, so it could persist this information somewhere to make this phase a no-op if the current compilation unit doesn't have any JUnit test. (And I'm not too worried about JUnit tests in non-test configuration, that's an edge-case).

Reply:
`needsJUnit4Fix` is only called for methods in traits, as a last resort of whether they need a mixin forwarder or not. Relying on this mechanism to store the information would miss a *lot* of methods, including `@Test` methods in `class`es, which are obviously the common case.